# NBA Draft Prediction

## Data Preprocessing

### Load the data

In [3]:
import pandas as pd

metadata = pd.read_csv('../data/raw/metadata.csv')
train_data = pd.read_csv('../data/raw//train.csv')
test_data = pd.read_csv('../data/raw//test.csv')

metadata

C:\Users\Sahil Kotak\AppData\Local\Temp\ipykernel_15748\1824229949.py:4: DtypeWarning: Columns (26) have mixed types. Specify dtype option on import or set low_memory=False.
  train_data = pd.read_csv('../data/raw//train.csv')


feature       name                                        description
0         1       team                                       Name of team
1         2       conf                                 Name of conference
2         3         GP                                       Games played
3         4    Min_per  Player's percentage of available team minutes ...
4         5       ORtg  ORtg - Offensive Rating (available since the 1...
..      ...        ...                                                ...
59       66        stl  STL - Steals (available since the 1973-74 seas...
60       67        blk  BLK - Blocks (available since the 1973-74 seas...
61       68        pts                                       PTS - Points
62       69  player_id                        Unique identifier of player
63       70    drafted  Target - Was the player drafted at the end of ...

[64 rows x 3 columns]

In [4]:
train_data

team  conf  GP  Min_per   Ortg   usg    eFG  TS_per   
0         South Alabama    SB  26     29.5   97.3  16.6   42.5   44.43  \
1              Utah St.   WAC  34     60.9  108.3  14.9   52.4   54.48   
2         South Florida    BE  27     72.0   96.2  21.8   45.7   47.98   
3            Pepperdine   WCC  30     44.5   97.7  16.0   53.6   53.69   
4               Pacific    BW  33     56.2   96.5  22.0   52.8   54.31   
...                 ...   ...  ..      ...    ...   ...    ...     ...   
56086           Niagara  MAAC   1      0.1    0.0  48.9    0.0    0.00   
56087  Northwestern St.  Slnd   1      0.2  206.9  35.9  100.0  102.56   
56088    Texas Southern  SWAC   1      0.6   48.5  28.9    0.0   52.63   
56089        Vanderbilt   SEC   1      0.1  300.0  20.0  150.0  150.00   
56090       Chicago St.   WAC  19     21.9   55.0  16.4   24.5   30.60   

       ORB_per  DRB_per  ...      dgbpm    oreb    dreb    treb     ast   
0          1.6      4.6  ...  -1.941150  0.1923  0.6154  0.8077  1.1923  \
1          3.8      6.3  ...  -0.247934  0.6765  1.2647  1.9412  1.8235   
2          2.1      8.0  ...  -0.883163  0.6296  2.3333  2.9630  1.9630   
3          4.1      9.4  ...  -0.393459  0.7000  1.4333  2.1333  1.1000   
4          8.3     18.6  ...  -0.668318  1.4242  3.3030  4.7273  0.8485   
...        ...      ...  ...        ...     ...     ...     ...     ...   
56086      0.0      0.0  ... -17.439600  0.0000  0.0000  0.0000  0.0000   
56087      0.0      0.0  ...   9.392350  0.0000  0.0000  0.0000  0.0000   
56088      0.0     15.9  ...  -3.240610  0.0000  1.0000  1.0000  0.0000   
56089      0.0      0.0  ...  16.362500  0.0000  0.0000  0.0000  0.0000   
56090      0.6      7.2  ...  -4.935520  0.0455  0.9091  0.9545  0.5909   

          stl     blk      pts                             player_id  drafted  
0      0.3462  0.0385   3.8846  7be2aead-da4e-4d13-a74b-4c1e692e2368      0.0  
1      0.4118  0.2353   5.9412  61de55d9-1582-4ea4-b593-44f6aa6524a6      0.0  
2      0.4815  0.0000  12.1852  efdc4cfc-9dd0-4bf8-acef-7273e4d5b655      0.0  
3      0.5667  0.1333   4.9333  14f05660-bb3c-4868-b3dd-09bcdb64279d      0.0  
4      0.4545  0.3333   7.5758  a58db52f-fbba-4e7b-83d0-371efcfed039      0.0  
...       ...     ...      ...                                   ...      ...  
56086  0.0000  0.0000   0.0000  9eded9ee-0eb4-49f4-914a-f58924797bdf      0.0  
56087  0.0000  0.0000   4.0000  5b539feb-1736-44ed-ba62-82bce86b12d6      0.0  
56088  0.0000  0.0000   2.0000  e95d08eb-1273-4094-9173-c47368222c0d      0.0  
56089  0.0000  0.0000   3.0000  2f315a7b-2e82-44a2-8597-1779102ace09      0.0  
56090  0.4091  0.0000   2.4091  65571c18-b2b4-4ba6-84db-7357a76e6f59      0.0  

[56091 rows x 64 columns]

In [5]:
train_data.shape

(56091, 64)

In [6]:
train_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 56091 entries, 0 to 56090
Data columns (total 64 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   team                 56091 non-null  object 
 1   conf                 56091 non-null  object 
 2   GP                   56091 non-null  int64  
 3   Min_per              56091 non-null  float64
 4   Ortg                 56091 non-null  float64
 5   usg                  56091 non-null  float64
 6   eFG                  56091 non-null  float64
 7   TS_per               56091 non-null  float64
 8   ORB_per              56091 non-null  float64
 9   DRB_per              56091 non-null  float64
 10  AST_per              56091 non-null  float64
 11  TO_per               56091 non-null  float64
 12  FTM                  56091 non-null  int64  
 13  FTA                  56091 non-null  int64  
 14  FT_per               56091 non-null  float64
 15  twoPM                56091 non-null 

In [7]:
train_data.describe(include="all")

team   conf            GP       Min_per          Ortg           usg   
count   56091  56091  56091.000000  56091.000000  56091.000000  56091.000000  \
unique    358     36           NaN           NaN           NaN           NaN   
top      Army    ACC           NaN           NaN           NaN           NaN   
freq      209   2297           NaN           NaN           NaN           NaN   
mean      NaN    NaN     23.411973     37.325229     91.858295     18.149265   
std       NaN    NaN     10.090542     28.061897     30.538819      6.204489   
min       NaN    NaN      1.000000      0.000000      0.000000      0.000000   
25%       NaN    NaN     16.000000      9.500000     84.100000     14.500000   
50%       NaN    NaN     28.000000     35.900000     97.100000     18.100000   
75%       NaN    NaN     31.000000     62.200000    106.900000     21.800000   
max       NaN    NaN     41.000000     98.000000    834.000000     50.000000   

                 eFG        TS_per       ORB_per       DRB_per  ...   
count   56091.000000  56091.000000  56091.000000  56091.000000  ...  \
unique           NaN           NaN           NaN           NaN  ...   
top              NaN           NaN           NaN           NaN  ...   
freq             NaN           NaN           NaN           NaN  ...   
mean       44.575513     47.676882      5.575376     12.673716  ...   
std        18.213813     17.414133      9.480239     10.878099  ...   
min         0.000000      0.000000      0.000000      0.000000  ...   
25%        40.100000     43.930000      1.800000      8.400000  ...   
50%        47.700000     50.850000      4.300000     11.900000  ...   
75%        53.200000     56.060000      8.100000     16.000000  ...   
max       150.000000    150.000000   1576.600000   1385.000000  ...   

               dgbpm          oreb          dreb          treb           ast   
count   56047.000000  56053.000000  56053.000000  56053.000000  56053.000000  \
unique           NaN           NaN           NaN           NaN           NaN   
top              NaN           NaN           NaN           NaN           NaN   
freq             NaN           NaN           NaN           NaN           NaN   
mean       -0.442297      0.770611      1.892593      2.663204      1.073038   
std         3.279438      0.738899      1.479148      2.099913      1.171148   
min      -100.984000      0.000000      0.000000      0.000000      0.000000   
25%        -1.728030      0.225800      0.700000      1.000000      0.222200   
50%        -0.313051      0.545500      1.633300      2.259300      0.677400   
75%         1.066415      1.111100      2.758600      3.833300      1.533300   
max        78.985500      5.933300     11.545500     14.533300     10.000000   

                 stl           blk           pts   
count   56053.000000  56053.000000  56053.000000  \
unique           NaN           NaN           NaN   
top              NaN           NaN           NaN   
freq             NaN           NaN           NaN   
mean        0.529475      0.282825      5.775381   
std         0.469152      0.417754      4.953723   
min         0.000000      0.000000      0.000000   
25%         0.160000      0.000000      1.571400   
50%         0.424200      0.133300      4.482800   
75%         0.793100      0.366700      9.069000   
max         4.000000      5.258100     30.090900   

                                   player_id       drafted  
count                                  56091  56091.000000  
unique                                 23929           NaN  
top     06dc8dc2-888e-4941-9106-4798cddfb9d2           NaN  
freq                                       6           NaN  
mean                                     NaN      0.009556  
std                                      NaN      0.097287  
min                                      NaN      0.000000  
25%                                      NaN      0.000000  
50%                                      NaN      0.000000  

## Data Cleaning

In [8]:
# Check for missing values in the dataset
missing_values = train_data.isnull().sum()

# Display columns with missing values and their counts
missing_values[missing_values > 0]

yr                       292
ht                        98
num                     4690
Rec_Rank               39055
ast_tov                 4190
rimmade                 6081
rimmade_rimmiss         6081
midmade                 6081
midmade_midmiss         6081
rim_ratio               9464
mid_ratio               9688
dunksmade               6081
dunksmiss_dunksmade     6081
dunks_ratio            30793
pick                   54705
drtg                      44
adrtg                     44
dporpag                   44
stops                     44
bpm                       44
obpm                      44
dbpm                      44
gbpm                      44
mp                        38
ogbpm                     44
dgbpm                     44
oreb                      38
dreb                      38
treb                      38
ast                       38
stl                       38
blk                       38
pts                       38
dtype: int64

In [9]:
# Calculate the percentage of missing values for each column
missing_percentage = (missing_values / len(train_data)) * 100

# Display columns with missing value percentages
missing_percentage[missing_percentage > 0].sort_values(ascending=False)

pick                   97.529015
Rec_Rank               69.627926
dunks_ratio            54.898290
mid_ratio              17.271933
rim_ratio              16.872582
rimmade                10.841311
rimmade_rimmiss        10.841311
midmade                10.841311
midmade_midmiss        10.841311
dunksmade              10.841311
dunksmiss_dunksmade    10.841311
num                     8.361413
ast_tov                 7.470004
yr                      0.520583
ht                      0.174716
obpm                    0.078444
dgbpm                   0.078444
ogbpm                   0.078444
gbpm                    0.078444
dbpm                    0.078444
adrtg                   0.078444
bpm                     0.078444
stops                   0.078444
dporpag                 0.078444
drtg                    0.078444
mp                      0.067747
oreb                    0.067747
dreb                    0.067747
treb                    0.067747
ast                     0.067747
stl       

Based on the percentage of missing values, we can make the following decisions:

Drop Columns: We can drop columns pick and Rec_Rank as they have a high percentage of missing values (97.53% and 69.63% respectively). The dunks_ratio column also has over 50% missing values, so we can consider dropping it as well.

Impute Values: For columns with a lower percentage of missing values, we can impute these values. For numerical columns, we'll use the median, and for categorical columns (like yr and ht), we'll use the mode.

Drop Rows: Since the percentage of missing values for some columns is very low (less than 0.1%), we can drop those rows.

### Handle non-numeric values

In [10]:
# Handle non-numeric values in 'num' column
median_value = pd.to_numeric(train_data['num'], errors='coerce').median()
train_data['num'] = pd.to_numeric(train_data['num'], errors='coerce').fillna(median_value)
test_data['num'] = pd.to_numeric(test_data['num'], errors='coerce').fillna(median_value)

# Convert 'yr' column to ordinal numbers
year_mapping = {'Fr': 1, 'So': 2, 'Jr': 3, 'Sr': 4}
train_data['yr'] = train_data['yr'].map(year_mapping)
test_data['yr'] = test_data['yr'].map(year_mapping)

### Handle missing values

In [11]:
# Drop columns with more than 50% missing data
columns_to_drop = ['pick', 'Rec_Rank', 'dunks_ratio', 'ht']
train_data.drop(columns_to_drop, axis=1, inplace=True)
test_data.drop(columns_to_drop, axis=1, inplace=True)

# Impute missing values with median for remaining columns with missing data
columns_to_impute = ['mid_ratio', 'rim_ratio', 'rimmade', 'rimmade_rimmiss', 'midmade', 'midmade_midmiss',
                     'dunksmade', 'dunksmiss_dunksmade', 'num', 'ast_tov', 'yr', 'obpm', 'dgbpm',
                     'ogbpm', 'gbpm', 'dbpm', 'adrtg', 'bpm', 'stops', 'dporpag', 'drtg', 'mp', 'oreb',
                     'dreb', 'treb', 'ast', 'stl', 'blk', 'pts']

for column in columns_to_impute:
    median_value = train_data[column].median()
    train_data[column].fillna(median_value, inplace=True)
    test_data[column].fillna(median_value, inplace=True)

## Model Training

In [4]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import roc_auc_score

# Split the data into train and validation sets
X = train_data.drop(['player_id', 'drafted'], axis=1)
y = train_data['drafted']
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=42)

# Identify numerical and categorical columns
numerical_cols = X_train.select_dtypes(include=['int64', 'float64']).columns.tolist()
categorical_cols = X_train.select_dtypes(include=['object']).columns.tolist()

# Preprocessing for numerical data
numerical_transformer = StandardScaler()

# Preprocessing for categorical data
categorical_transformer = OneHotEncoder(handle_unknown='ignore')

# Bundle preprocessing for numerical and categorical data
preprocessor = ColumnTransformer(
    transformers=[
        ('num', numerical_transformer, numerical_cols),
        ('cat', categorical_transformer, categorical_cols)
    ])

# Define the model
model = LogisticRegression(max_iter=1000, random_state=42)

# Bundle preprocessing and modeling code in a pipeline
pipeline = Pipeline(steps=[('preprocessor', preprocessor),
                           ('model', model)])

# Preprocessing of training data, train model
pipeline.fit(X_train, y_train)

# Preprocessing of validation data, get predictions
val_preds = pipeline.predict_proba(X_val)[:, 1]

# Get the AUROC score
val_score = roc_auc_score(y_val, val_preds)
val_score

0.9910943972095052

## Generate Predictions

In [6]:
# Make predictions on the test data
test_preds = pipeline.predict_proba(test_data.drop('player_id', axis=1))[:, 1]

# Create a submission DataFrame
submission = pd.DataFrame({'player_id': test_data['player_id'], 'drafted': test_preds})

# Save the submission DataFrame to a CSV file
submission_path = 'submissions_logisticregression1.csv'
submission.to_csv(submission_path, index=False)
submission_path

'submissions_logisticregression1.csv'